// gatsby-transformer-sharp/on-node-create.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

const supportedExtensions = {
  jpeg: true,
  jpg: true,
  png: true,
  webp: true,
  tif: true,
  tiff: true
};

module.exports =
/*#__PURE__*/
function () {
  var _onCreateNode = (0, _asyncToGenerator2.default)(function* ({
    node,
    actions,
    createNodeId
  }) {
    const createNode = actions.createNode,
          createParentChildLink = actions.createParentChildLink;

    if (!supportedExtensions[node.extension]) {
      return;
    }

    const imageNode = {
      id: createNodeId(`${node.id} >> ImageSharp`),
      children: [],
      parent: node.id,
      internal: {
        contentDigest: `${node.internal.contentDigest}`,
        type: `ImageSharp`
      }
    };
    createNode(imageNode);
    createParentChildLink({
      parent: node,
      child: imageNode
    });
    return;
  });

  return function onCreateNode(_x) {
    return _onCreateNode.apply(this, arguments);
  };
}();
